core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -522,14 +657,20 @@ def hash143_preimage_hash(self, txi: TxInputType, pubkeyhash: bytes) -> bytes:
         writers.write_uint32(h_preimage, self.tx.lock_time)
 
         # nHashType
-        writers.write_uint32(h_preimage, self.get_hash_type())
+        writers.write_uint32(h_preimage, self.get_sighash_type())
 
         return writers.get_tx_hash(h_preimage, double=self.coin.sign_hash_double)
 
 
 def input_is_segwit(txi: TxInputType) -> bool:
-    return txi.script_type in helpers.SEGWIT_INPUT_SCRIPT_TYPES
+    return txi.script_type in common.SEGWIT_INPUT_SCRIPT_TYPES or (
+        txi.script_type == InputScriptType.EXTERNAL and txi.witness is not None
+    )
 
 
 def input_is_nonsegwit(txi: TxInputType) -> bool:
-    return txi.script_type in helpers.NONSEGWIT_INPUT_SCRIPT_TYPES
+    return txi.script_type in common.NONSEGWIT_INPUT_SCRIPT_TYPES
```

Comment:
for symmetry this should also include external inputs, or be renamed

Reply:
f305743a9f055d764089ec59ac35c4d86caf9e90